python/dgl/nn/pytorch/conv/cugraph_relgraphconv.py

```diff
@@ -0,0 +1,388 @@
+        agg_hg_basis_post_bwd_int32,
+        agg_hg_basis_post_bwd_int64,
+        agg_hg_basis_post_fwd_int32,
+        agg_hg_basis_post_fwd_int64,
+    )
+    from pylibcugraphops.legacy.structure.graph_types import (
+        message_flow_graph_hg_csr_int32,
+        message_flow_graph_hg_csr_int64,
+    )
+except ModuleNotFoundError:
+    raise ModuleNotFoundError(
+        "dgl.nn.CuGraphRelGraphConv requires pylibcugraphops to be installed."
+    )
+
+CUDA_SM_PER_BLOCK = 49152
+
+
+class RelGraphConvAgg(th.autograd.Function):
+    r"""Custom autograd function for R-GCN aggregation layer that uses the
+    aggregation functions in cugraph-ops."""
+
+    @staticmethod
+    def forward(ctx, g, fanout, num_rels, edge_types, feat, coeff):
+        r"""Compute the forward pass of R-GCN aggregation layer.
+
+        Parameters
+        ----------
+        g : DGLGraph
+            The graph.
+        fanout : int
+            Maximum in-degree of nodes.
+        num_rels : int
+            Number of relations.
+        edge_types : torch.Tensor
+            A 1D tensor of edge types. Note that pylibcugraphops only accepts
+            edge_types in int32, so any input of other integer types will be casted
+            into int32, thus introducing some overhead. Pass in int32 tensor for
+            best performance.
+        feat : torch.Tensor, dtype=torch.float32, requires_grad=True
+            Tensor of source node features. Shape: (num_src_nodes, in_feat).
+        coeff : torch.Tensor, dtype=torch.float32, requires_grad=True
+            The coefficient matrix used in basis-decomposition regularization.
+            Shape: (num_rels, num_bases). It should be set to ``None``
+            when ``regularizer=None``.
+
+        Returns
+        -------
+        agg_output : torch.Tensor, dtype=torch.float32
+            Aggregation output. Shape: (num_dst_nodes, num_rels * in_feat)
+            when ``regularizer=None``; Shape: (num_dst_nodes, num_bases * in_feat)
+            when ``regularizer='basis'``.
+        """
+        ctx.device = g.device
+        if g.idtype == th.int32:
+            mfg_csr_func = message_flow_graph_hg_csr_int32
+            agg_fwd_func = agg_hg_basis_post_fwd_int32
+        elif g.idtype == th.int64:
+            mfg_csr_func = message_flow_graph_hg_csr_int64
+            agg_fwd_func = agg_hg_basis_post_fwd_int64
+        else:
+            raise TypeError(
+                f"Supported ID type: torch.int32 or torch.int64, "
+                f"but got {g.idtype}"
+            )
+        ctx.idtype = g.idtype
+
+        _in_feat = feat.shape[-1]
+        indptr, indices, edge_ids = g.adj_sparse("csc")
+        # edge_ids is in a mixed order, need to permutate incoming etypes
+        # and stash the result for backward propagation
+        ctx.edge_types_int32 = edge_types[edge_ids.long()].int()
+        # node_types are not needed by the post-variant RGCN aggregation functions
```

Comment:
What does "post-variant" mean here?

Reply:
This was some notes related to the APIs in cugraph-ops. I have cleaned up the comments.